src/spring2024/s24_my_ptr.cpp

```diff
@@ -0,0 +1,250 @@
+#include <iostream>
+#include <memory>
+#include <utility>
+
+// This file contains codes used in the Spring2024 15445/645 C++ bootcamp.
+// It dives deeply into C++ new features like move constructor/assignment, move semantics, unique_ptr, shared_ptr,
+// wrapper class, etc., by implementing a simple version of unique_ptr from scratch.
+// **IMPORTANT**: please BEGIN reading from MAIN function!
+
+// Notice: it is just a simple version of std::unique_pointer<T>. The real implementation is much more complex!
+// A template allows us to replace any type T, with what we want later in our code.
+template <typename T>
+class Pointer {
+ public:
+  Pointer() {
+    ptr_ = new T;
+    *ptr_ = 0;
+    std::cout << "New object on the heap: " << *ptr_ << std::endl;
+  }
+  Pointer(T val) {
+    ptr_ = new T;
+    *ptr_ = val;
+    std::cout << "New object on the heap: " << val << std::endl;
+  }
+  // DESTRUCTOR: Be called whenever an instance gets out of scope (just when the stack pops).
+  ~Pointer() {
+    if (ptr_) {
+      std::cout << "Freed: " << *ptr_ << std::endl;
+      delete ptr_;
+    }
+  }
+
+  // Copy-init constructor is explicitely deleted.
+  Pointer(const Pointer<T> &) = delete;
+  // Copy-assign operator is explicitely deleted.
+  Pointer<T> &operator=(const Pointer<T> &) = delete;
+
+  // Add move constructor: useful when we need to EXTEND the lifetime of an object!
```

Comment:
(Is this file meant to be read after move_constructors.cpp? If so, this is fine. If not, maybe (consult this other file for the meaning of lifetime)

Reply:
Also this section should probably be read after the move semantics files given we go into move constructors and assignments here.

Reply:
Yeah! Good point! I add an important note in the beginning of this file.